### ISSUE TITLE
[Discussion] Rename Category to FiniteCategory, and add a Category class that allows infinite ones

### ISSUE THREAD

**Issue by sava-1729** [sympy/sympy#18983]
Hi,
From what I understood in the categories module, the ``Category`` class allows only finitely many objects, hence a finite category. However, the standard and most important categories such as the Category of Sets, Category of Monoids, Category of Vector Spaces, etc. are infinite. And I propose to implement these as a part of GSoC 2020 (I'll put up the formal proposal soon). But to implement any of these I will definitely need some sort of a base class, that represents abstract infinite categories. How should I go about implementing that?
In order to define an abstract category `C`, we need the following data:

- objects `a`, `b`, . . . ,

- for any two objects `a` and `b`, a set of morphisms
`C(a, b)`

- for each object `a`, a distinguished element
of `C(a, a)` denoted `id_a` .

- a composition rule `*` for morphisms, that is associative and unital

One way of implementing it, could be defining symbolically what the objects of a category are, its morphisms, id's etc. What I mean is something like:
```
class AbstractCategory(Basic):
    objects_type = None
    morphisms_type = None
    composition_rule = None
    pass #various util functions such as _contains will also be defined in this class

class CategoryOfSets(AbstractCategory):
    objects_type = Set
    morphisms_type = Function
    composition_rule = CompositeFunction
```
But there are a few problems in my approach:

- I couldn't find any suitable symbolic class to represent **set-theoretic functions**, i.e., functions that are aware of their domain and codomain (I looked at the `Function` class, but it didn't seem like one can specify domain and codomain). If there is a suitable class, please let me know, it would be a big help.

- This way, only the super-categories can be implemented, i.e., I don't see a direct way to derive something like `CategoryOfFiniteSets` from `CategoryOfSets`.

Suggestions/ideas/views are greatly appreciated.
Also, respected mentors, could you please tell me if this, in its entirety, make a valid project for GSoC 2020?

**Comment by sylee957** [comment]
I think that working on sympy Function architecture to achieve that is too much huge because functions are classes than objects.
But I see there are already `Morphism` in category which has zero connection with sympy function, but define in their own way to represent domain and codomain, 
So some work can begin independently in category module, and the task of connecting things sympy functions with there can be done later.

**Comment by sava-1729** [comment]
Sorry for my long absence.
I was wondering how to implement the set-theoretic `Function` object, which will be required to define the `CategoryOfSets`.
Initially I thought I could define it to be simply an object that contains the following information:

- `domain`, instance of sympy.sets.sets.Set

- `codomain`, instance of sympy.sets.sets.Set

- `mapping`, instance of sympy.sets.sets.Set

where `mapping` is a subset of `ProductSet(domain, codomain)`, such that for every element `x` in `domain`, there exists a unique element `M` in `mapping` such that `M = (x, y)` for some `y` in `codomain`
But I don't know how to check the above condition given the three objects `domain`, `codomain`, `mapping`? Any ideas/suggestions/views regarding the same are highly appreciated!

**Comment by asmeurer** [comment]
I agree with the basic principle of this, which is that we should support infinite categories. 

Can you clarify how your class structure would work as a Basic object? All Basic objects should store all the information about them in their `.args`, so that they can be recreated with `obj.func(*obj.args)`. What would be the .args for an AbstractCategory object?

**Comment by sava-1729** [comment]
So I've been thinking about that. And I realized that we probably need not allow users to instantiate the `AbstractCategory` class itself, because 

1. We can't explicitly store all objects, morphisms, etc of a general category (as there could be infinitely many)

2. Even if we figure out some way to symbolically represent a category that can store infinitely many objects, it would be using some sort of a **user-defined rule** (the principle on which sympy.sets.facnysets.ImageSet is defined). But then even to define very simple finite categories, the user would need to define some rule, which seems inconvenient.

So my suggestion is that the `AbstractCategory` class would be there just to have an overseeing parent, users would not be allowed to instantiate that.

Now, to address the main issue: How to support infinite categories?

To support infinite sets in sympy, they were divided into various types like `ImageSet` , `ProductSet` etc. and important sets have been explicitly laid out, like Naturals, Reals, Integers, etc.
Similarly in order to support infinite categories, there should be some symbolic way to store an infinite number of objects (of the category) in a **Class** (to avoid confusion, I have used **class** to denote a python class and **Class** to denote `sympy.categories.baseclasses.Class`, everywhere). So we need to define an analogue of `ImageSet` for **Class**


The need to define that analogue (as opposed to using `ImageSet` itself) is that strictly speaking, a Category is defined to contain a **Class** of objects and **not a set** of objects (because then, the category of sets would have to contain **the set of all sets**, which leads to the Russell's paradox - controversial territory). So a Class is defined in category theory just to get rid of all such discrepancies, and it is by definition just a collection of objects.

So we need to define an `ImageClass`, something that takes in a `Lambda` object (call it `filter`) and some instances of `Class`, where the `filter` decides which elements of `Class` will exist in the `ImageClass`. There can be multiple domain classes depending on the number of variables in `filter`, in the example below, I have assumed `filter` is a single variable function. It should essentially look something like this:
```
from sympy.core.function import Lambda

class ImageClass(Class):
    def __new__(cls, filter, domain_class):
        """
        filter is supposed to work like the following
        >>> A = Symbol("A")
        >>> #The Lambda object given below verifies whether a set is a real number set (subset of Reals) or not
        >>> lambda_filter = Lambda(A, A.is_subset(Reals))
        >>> lambda_filter(FiniteSet(1,2,3))
        True
        >>> lambda_filter(FiniteSet(Symbol('x'), Symbol('y'), Symbol('z')))
        False
        """
        assert isinstance(filter, Lambda)
        assert isinstance(domain_class, Class)
        return Basic.__new__(cls, filter, domain_class)

    filter = property(lambda self: self.args[0])
    domain_class = property(lambda self: self.args[1])

    def _contains(self, other):
        return (other in self.domain_class) and self.filter(other)
```
Of course, the slight issue in this is that currently, `Symbol` cannot represent a Set, I think it's not supported. So for that we can extend `Symbol`, or define a separate class `SetSymbol`.

So apart from having satisfied all the above prerequisites, also assume that we already have some **class** that symbolically represents the **Class of all sets** (or informally, the collection of all sets), let's denote that by `AllSets` (similar to S.Naturals)

Now, if I want to represent the **Class of all finite real number sets**, I could do:
```
>>> A = Symbol("A")
>>> fin_real_filter = Lambda(A, A.is_finite and A.is_subset(Reals))
>>> fin_real_sets = ImageClass(fin_real_filter, AllSets)
>>> S.Integers in fin_real_sets
False
>>> FiniteSet(1,2,3,4) in fin_real_sets
True
```
So clearly fin_real_sets represents a collection (formally, a **Class**) of all finite sets of real numbers. So this can now be used to form the category of all finite sets of real numbers. And to define such custom infinite categories we can define a separate class:
```
class CustomCategory(Basic):
    def __new__(cls, objects, morphisms):
        assert isinstance(objects, Class)
        assert isinstance(morphisms, Class)
        return Basic.__new__(cls, objects, morphisms)

>>> cat_fin_real_sets = CustomCategory(fin_real_sets, morphisms)
```
I'm not so sure how to deal with all/infinitely many `morphisms` though. Maybe that can also be achieved using **Class**.
@scolobb @asmeurer @sylee957 do you have any suggestions/ideas/thoughts?

**Comment by scolobb** [comment]
Hi @sava-1729 and thank you for your interest in this stuff!

I'll start with a disclaimer: as you have probably noticed, I haven't touched SymPy or the category theory module for ages now, so everything is a bit rusty in my brain. But, let's see if I can still be useful.

> So my suggestion is that the `AbstractCategory` class would be there just to have an overseeing parent, users would not be allowed to instantiate that.

To me this a question of "instance of a class" vs. "derived class".  I think (but don't trust me on this) my original intention was to define various objects of type `AbstractCategory` which would correspond to different categories.  Thus, the word "abstract" in there would not refer to the same thing as the word "abstract" in "abstract class" (I'm sure you realize that, I'm just saying to be 100% clear).  I'm not completely sure that making `AbstractCategory` an abstract class would bring more power, but perhaps I'm missing something.

> So we need to define an `ImageClass`

I completely agree.  The definition you wrote makes a lot of sense to me, but I'm no longer up-to-date with SymPy coding style.

> assume that we already have some class that symbolically represents the Class of all sets

I suppose that should be relatively easy to achieve by checking whether the argument of `filter` is a (SymPy) set.

> I'm not so sure how to deal with all/infinitely many morphisms though. Maybe that can also be achieved using Class.

That would be my idea.  Morphisms in abstract categories are just arrows between the objects, so the morphisms of the category could be represented as a subcollection of the product class Objects × Objects.  Now, technically, the morphisms between any two given objects should form a set (it depends on who you ask, but take away this property and some nice results like Yoneda lemma fall apart).  Let's give this problem a small think.

**Comment by scolobb** [comment]
OK @sava-1729 I've read the whole thread and now I have a better grip on the context.

As @sylee957 said, the CT module defines its own fundamental objects, so you don't need to connect `Morphism` to `Function`.  In fact, a morphism is more general than a function, so focusing on `Function` too much would not be very relevant (especially for abstract categories).  This is partly good news, because the CT module exists in a somewhat separate world.  Now, @asmeurer 's comment on making the connection between `Class` and `Basic` explicit is crucially important: the CT module has to stay part of SymPy, so it has to play nicely with its objects.

Anyway, thanks again for your interest and do feel free to ping me (if you don't I'll probably miss your message).
